Proposal for class definitions

ch-sa opened the pull request:
Hi @chingyulin,

unfortunately I cannot put you as reviewer but I would like to get your opinion on this idea:

I thought it makes sense to have the label definitions/ schema in an extra single file as you suggested.
This makes also sense for the object detection use case as label schemas are mostly project specific and thus should be located close to the labels. Also it would be a cool feature to have label colors for object detection.

So what I did is adding more data to the schema file and name + place it more generic so it fits both cases.

What do you think of this idea? (This is just a draft, I would continue with the implement then ...)

Changes:
 - use structure with all details
 - includes id, color and default flag
 - could be autogenerated and partly optional
 - would be reusable for object detection

chingyulin commented:
Hi, 

I think it is a good idea. Perhaps the `default` can be outside `classes` as it can only have one class. I also like the idea that sharing the features of label coloring for segmentation and object detection.  :)

chingyulin commented:
> Btw. is your other pull request already in a state were I should start reviewing?

Not yet, I've been busy on other things. Will try to make it this weekend or next.

ch-sa commented:
I put the loading of the `_classes.json` file with the class definitions into the label io.

That means whenever the label folder changes we can reload the file and with every new label we can update the definitions vice versa.

My idea is to show a dialog on startup which offers the user to define these class names and select colors or if the file exist to simply check the config.

BBox coloring is already working for existing classes. Still need to implement it for newly created classes though ...

chingyulin commented:
I think it's a good idea to have a start up dialog to set the label definition.

ch-sa commented:
Still WIP, but ...

1. I rebased the feature branch with master and force pushed
2. rebased this on it
3. added a short proof of context to change the class color via a context menu on the labels

![color_change](https://user-images.githubusercontent.com/26072175/188332048-c6baeb43-189e-45cc-a094-d6bc3e5a1f30.gif)

What do you think?

Obviously should open up a color selection dialog and not change them randomly ;)

chingyulin commented:
I think this functionality looks great. UX-wise, I wonder if we should make those information directly visible in stead of showing it only if the user right clicks it? Also it seems two of the three options in the menu are configurable in the original design `Current Class` edit bar and the `Delete button`? Or is the intention to remove those?

ch-sa commented:
> I think this functionality looks great. UX-wise, I wonder if we should make those information directly visible in stead of showing it only if the user right clicks it? Also it seems two of the three options in the menu are configurable in the original design `Current Class` edit bar and the `Delete button`? Or is the intention to remove those?

@chingyulin  Yes, you are right, that the deletion and renaming is duplicated and probably does not make sense.

I think the context menu mostly makes sense for options that are probably not needed that often, like:
1. Change label color
2. Make label default for new labels

ch-sa commented:
OK, the label definitions slowly become more concrete:

* The LabelConfig is now a new Singleton independent of other classes
* This makes using it much simpler cause you get it via `LabelConfig()` everywhere
* I don't think the mode classes (bbox, pcd) need a `color` attribute themselves as it can always be derived
  * maybe for segmentation we need a private cache though (?)

There are still some construction sites and I still have to get the segmentation to work with it. But I think the overall idea is clear now.

I general I wonder a bit if we really should go with two separate configuration files, `_classes.json` and `config.ini` or if they could be unified for example in a `project.yaml`.

ch-sa commented:
@chingyulin Segmentation works again ...

But some things were a bit difficult to understand for me and I think there is one general problem:
If I understand the code correctly the translation of labels to colors is based on the relative position of a class within the class list and not the actual `class.id,` or?

I think we can simplify a bit there ... I will comment in the code.

ch-sa commented on labelCloud/model/point_cloud.py:
```diff
@@ -48,7 +48,6 @@ def __init__(
         self,
         path: Path,
         points: npt.NDArray[np.float32],
-        label_definition: Dict[str, int],
```
This is no longer necessary as it is directly retrievable from `LabelConfig`.

ch-sa commented on labelCloud/model/point_cloud.py:
```diff
@@ -59,12 +58,11 @@ def __init__(
         if self.SEGMENTATION:
             self.labels = segmentation_labels
-            self.label_color_map = get_distinct_colors(len(label_definition))
+            self.label_color_map = LabelConfig().get_color_map()
```
You expect colors with 3 values (RGB). Currently LabelConfig also supports alpha = RGBA. We should decide if that makes sense.

Can also just drop that support.

ch-sa replied:
I know changed to RGB only. I think for class colors alpha does not makes sense.

ch-sa commented on labelCloud/model/point_cloud.py:
```diff
@@ -132,7 +130,7 @@ def label_colors(self) -> npt.NDArray[np.float32]:
         if self.labels is not None:
-            label_one_hot = np.eye(len(self.label_definition))[self.labels]
+            label_one_hot = np.eye(LabelConfig().nb_of_classes)[self.labels]
             colors = np.dot(label_one_hot, self.label_color_map).astype(np.float32)  # type: ignore
```
So it does not work with gaps in between the ids (a) and if the id sequence differs from the list sequence (b).

However, I was thinking that we don't even open up the ids for the user and set them increasing and in order automatically. That would then only be the problem if the user edits the `_classes.json` manually.

chingyulin replied:
Ah, yes it's a logic flaw in the code. I'll make a fix soon.

chingyulin replied:
I have a PR here to fix the problem https://github.com/ch-sa/labelCloud/pull/107

We could also remove the ids in _classes.json but I feel some users would want the flexibility to define them while maintain the possibility to extend it. For example, 
Some might want {"unassigned": 0, "label1": 1, "label2": 2, "clutter": 255} and later on want to change it to {"unassigned": 0, "label1": 1, "label2": 2,  , "label3": 3, "clutter": 255}. This is easier to do with id field in place. What do you think?

ch-sa replied:
Yes, also renaming of class names might be easier with ids (if we would use the id as the class reference later).

I will have a look at your PR now.

ch-sa commented:
I finished the color dialog and I think the `LabelConfig()` class is now final.

(I also noticed that to allow users the renaming of classes would require a refactoring that only references labels via the id and not the name. But this is a topic for future iterations. :wink: )

chingyulin reviewed (commented):
The unification of the label definitions across two modes looks good to me. A bit confused about the startup dialog - what did you implement? I see some changes in actions but couldn't see the startup diaglog

ch-sa commented on labelCloud/io/labels/config.py:
```diff
@@ -0,0 +1,105 @@
+from ...utils.color import hex_to_rgb, rgb_to_hex
+from ...utils.singleton import SingletonABCMeta
+
+CONFIG_FILE = "_classes.json"  # TODO: Move to config?
```
I see, wasn't used so I drop it.

ch-sa commented on labelCloud/model/bbox.py:
```diff
@@ -153,17 +162,17 @@ def draw_bbox(self, highlighted: bool = False) -> None:
-        if highlighted:
-            bbox_color = (0, 1, 0, 1)
+        bbox_color = LabelConfig().get_class_color(self.classname)
+        # if highlighted:
+        #     bbox_color = (0, 1, 0, 1)
```
Good point. I only commented for testing. Will activate again.

I think it should still be clear which box is the currently active one by this lime green color.

ch-sa commented:
> The unification of the label definitions across two modes looks good to me. A bit confused about the startup dialog - what did you implement? I see some changes in actions but couldn't see the startup diaglog

@chingyulin I did not add the start up dialog. This PR was only about adding the new class structure and get everything clean about.

The start up dialog would now be next and (hopefully) last thing necessary before we can merge with master.

This PR is now ready to be merged as soon as you approve :)

chingyulin commented:
Looks good to me :)

chingyulin reviewed (approved):
Looks good to me :)
